be/src/exec/tablet_sink.cpp

```diff
@@ -228,23 +228,33 @@ Status NodeChannel::_serialize_chunk(const vectorized::Chunk* src, ChunkPB* dst)
     if (_compress_codec != nullptr && uncompressed_size > 0) {
         SCOPED_TIMER(_parent->_compress_timer);
 
-        // Try compressing data to _compression_scratch, swap if compressed data is smaller
-        int max_compressed_size = _compress_codec->max_compressed_len(uncompressed_size);
+        if (_compress_codec->type() == CompressionTypePB::LZ4_FRAME ||
+            _compress_codec->type() == CompressionTypePB::ZSTD || _compress_codec->type() == CompressionTypePB::LZ4) {
+            Slice compressed_slice;
+            Slice input(dst->data());
+            _compress_codec->compress(input, &compressed_slice, true, uncompressed_size, nullptr,
+                                      &_compression_scratch);
+        } else {
+            int max_compressed_size = _compress_codec->max_compressed_len(uncompressed_size);
+
+            if (_compression_scratch.size() < max_compressed_size) {
+                _compression_scratch.resize(max_compressed_size);
```

Comment:
_compress_scratch may be resized multiple times and it also allocates a large memory when resizing happens. In most cases, the chunks' size is similar if they have similar rows, so could we take a more aggressive strategy?such as allocate a large enough memory at the first time to avoid the following allocation.

Reply:
the _compression_scratch here will be swaped out each time

```c++
dst->mutable_data()->swap(reinterpret_cast<std::string&>(_compression_scratch));
```

Reply:
I did something same  in `CompressContext`, each compression context has a `compression_buffer`, which is shared by all users who use the same CompressContext. 
With this, for LZ4/LZ4F/ZSTD compression,  `_compress_scratch ` will only allocate the exact compressed size (not a size which is much larger than the compressed size)

```c++
struct LZ4CompressContext {
    LZ4CompressContext() : ctx(nullptr), compression_fail(false), compression_count(0) {}

    // LZ4 compression context
    LZ4_stream_t* ctx;

    bool compression_fail;
    uint32_t compression_count;

    // This compression_buffer is shared under the same compress context.
    // We will first save the compression result on this compression_buffer,
    // and then copy the result in compression_buffer to the target output `compressed_body`.
    // Through this way, we can avoid allocate a very large compressed_body first and then shrink it.
    faststring compression_buffer;
};

```